Frontend scaling

pstibrany opened the pull request:
**What this PR does**:
This PR implements scaleable query-frontend by extracting queue (called "scheduler") into separate component. Implementation doesn't change existing query-frontend, but introduces completely new one (copying the old one). Followup PR will refactor both implementations to reuse shared code.

Current thinking is that there will be no migration path from old to new model, but old model (no scheduler) will be available and used in single-binary mode, or with configuration without scheduler. Scheduler will not be part of single-binary, but can run separately and frontend and queriers in single-binary mode can then be configured to use it.

When using query-frontend V2, query-scheduler and new query workers, flow of the request is as follows:

1. Query arrives to query-frontend.
2. Query is forwarded to random query-sheduler.
3. Random querier will pick the query from query-sheduler.
4. When querier is finished with processing of the query, it sends response back directly query-frontend, avoiding the scheduler.

Query-scheduler exposes two gRPC interfaces: `SchedulerForFrontend` and `SchedulerForQuerier`. Both have single "loop" method which follows the same pattern as original frontend does – exchanging requests/responses in lockstep.

When scheduler finds that frontend has completely disconnected (which it can do thanks to "loop" method), it will cancel all requests from given frontend. Frontend can also cancel request that is already enqueued in scheduler (or even running in the querier). It does so by sending cancellation request to scheduler. Cancelling request that is already running on querier means closing the "loop" method (by scheduler) for given querier.

Both frontend and querier-worker use DNS watcher to monitor what schedulers are available, and connect to all of them. Thanks to using schedulers as the middleman, we can now scale up query-frontends much more easily. Number of schedulers should be low (eg. two, for redudancy).

This PR supercedes https://github.com/cortexproject/cortex/pull/3255 (while taking parts of it)

This PR also splits old `Frontend` into `Handler` (HTTP handler that uses generic RoundTripper), and `GrpcRoundTripper` + adapter (both old and new Frontends implement `GrpcRoundTripper` interface). V1 and V2 frontend selection is based on whether scheduler address is configured.

**Which issue(s) this PR fixes**:
Fixes #1150 

**Notes to reviewer**: I'd start with `frontend2/frontend.proto` to make sure it makes sense. 

TODO:
- More tests
- Documentation
- Refactor frontend and frontend2 package to reuse common code (follow-up PR):
  - move dns_watcher.go from frontend2 to frontend, and refactor querier worker to use it
  - querier workers are quite similar, but in v1 querier gets queries from frontend and sends response back to frontend, while in v2 querier gets queries from scheduler and sends response directly to frontend (not scheduler). Ideally this should be refactored to reuse common code.
  - `frontend_querier_queues.go` is in frontend2 just a copy with different request type. If we use `interface{}`, we don't need two copies.

**Checklist**
- [x] Tests updated
- [ ] Documentation added
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

pstibrany commented:
Marking as ready for review, as I don't expect big changes at this point, even though I still plan to add more tests to new frontend2 code.

pstibrany commented:
I've made this diagram to help in the review.

<img width="1436" alt="frontend-scaling-diagram" src="https://user-images.githubusercontent.com/895919/97309327-274ab080-1862-11eb-8e96-499d3ef5ac66.png">

Note that part of this PR was extracting `frontend.grpcRoundTripperAdapter` and `frontend.Handler` from existing `frontend.Frontend`, to make it reusable. Most changes in original `frontend` in this PR are related to this, and keeping configuration compatibility.

Main flow from the diagram is in frontend2 package.

pstibrany commented on pkg/querier/frontend2/frontend.proto:
```diff
@@ -0,0 +1,94 @@
+    //
+    // Long-running loop is used to detect broken connection between frontend and scheduler. This is important for both
+    // parties... if connection breaks, frontend can cancel (and possibly retry on different scheduler) all pending
+    // requests sent to this scheduler, while scheduler can cancel queued requests from given frontend.
```
When last connection from frontend disconnects (ie. last FrontendLoop call for given frontend-address exits), scheduler will cancel all requests from given frontend. It does so by cancelling frontend-specific context.

Apart from that, the only other way how request from frontend is cancelled is when frontend itself sends cancellation request to scheduler.

pstibrany commented on pkg/querier/frontend2/frontend.proto:
```diff
@@ -0,0 +1,94 @@
+    // Used by INIT message. Will be put into all requests passed to querier.
+    string frontendAddress = 2;
+
+    // Used by ENQUEUE and CANCEL. Each enqueued query must have queryID higher than previous one.
```
queryID is unique per frontend.

Different frontends can use same queryIDs for different queries.

(Since each frontend starts with a random uint64 number this is unlikely, but reason for doing randomization is to avoid conflicts when frontend restarts -- it could receive response from querier for wrong query).

pstibrany commented on pkg/cortex/cortex.go:
```diff
@@ -293,11 +296,15 @@ func New(cfg Config) (*Cortex, error) {
-		[]string{"/cortex.Ingester/TransferChunks", "/frontend.Frontend/Process"})
+		// Also don't check auth for these gRPC methods, since single call is used for multiple users (or no user like health check).
+		[]string{
+			"/grpc.health.v1.Health/Check",
```
Client pool uses health check to remote stale clients from the pool. When issuing Check method, it sets user "0", to avoid authentication errors. (https://github.com/cortexproject/cortex/blob/4b40f20aa8f0d0bead9e65e6a25fdee78a8787a3/pkg/ring/client/pool.go#L193)

I've tried to remove that setting of "0" and add this exception instead. It worked fine between querier and frontend, but not between querier and ingester, simply because I haven't deployed updated ingester during my testing. I have then realized that it would be bit painful to migrate if I had removed that setting of "0" user, so first step in migration is just adding this exception.

That said, we can remove it from this PR as it's currently not needed.

pstibrany commented on pkg/cortex/modules.go:
```diff
@@ -506,18 +503,36 @@ func (t *Cortex) initQueryFrontendTripperware() (serv services.Service, err erro
+
+	t.API.RegisterQueryFrontendHandler(handler)
+
+	if frontendV1 != nil {
```
No, that's not possible.

Both query-frontend and querier must scheduler or neither should.

Migration from setup with query-frontend and querier but no scheduler is possible by:

1) starting scheduler (or two)
2) starting query-frontend and querier configured with scheduler
3) redirecting queries from old query-frontend (without scheduler) to new query-frontend
4) once no traffic is coming to old query-frontends, they can be stopped, together with old queriers.

It would be possible with some extra effort to provide migration path at query-frontend level, but this PR doesn't do that.

ranton256 replied:
Okay, so basically the migration path would be spin up new queriers + new query frontends + scheduler, then switch traffic from old query frontends and old queriers, then turn down old queriers and old query frontends. That seems like it will work, just requires running more components during transition.

pstibrany commented on pkg/querier/frontend/handler.go:
```diff
@@ -0,0 +1,153 @@
+	}
+}
+
+//// Handler creates handler for HTTP requests.
```
No, just oversight :) Thanks.

pstibrany commented on pkg/querier/frontend2/frontend2.go:
```diff
@@ -0,0 +1,290 @@
+	f.requests.put(freq)
+	defer f.requests.delete(freq.queryID)
+
+	retries := f.cfg.WorkerConcurrency + 1 // To make sure we hit at least two different schedulers.
```
You mean number of retries?

The worst that can possibly happen is that no worker will take this request, and request gets stuck until context is cancelled. This can happen with 0 retries (eg. if no scheduler was found).

My thinking behind this value is:

- IF there are two schedulers,
- and one is shutting down
- and request is processed by all worker-goroutines for this scheduler (= WorkerConcurrency), 

then we still hit at least one worker for other scheduler.

pstibrany commented on pkg/querier/frontend2/frontend2.go:
```diff
@@ -0,0 +1,290 @@
+		if er.retry {
+			retries--
+			if retries > 0 {
+				goto enqueueAgain
```
There is no backoff here. We're trying to handoff request to some worker-goroutine to forward it to scheduler. If this fails with one, we want to try another one as soon as possible with no waiting.

Note that if all workers are busy handling other requests, we don't get to this point at all -- that would block when sending request to channel. We only get here if some worker was able to receive the request via channel, but forwarding it to scheduler has failed.

pstibrany commented on pkg/querier/frontend2/frontend2.go:
```diff
@@ -0,0 +1,290 @@
+		return nil
+	}
+
+	msg := fmt.Sprintf("not ready: number of schedulers this worker is connected to is %d", workers)
```
You're not missing anything. I had the same comment when this was introduced in original frontend package: https://github.com/cortexproject/cortex/pull/2733#discussion_r447795688

Joe claimed it might be worth just in case it gets negative at some point. I still don't see much value in it, but since it's harmless, I left it as is.

pstibrany commented on pkg/querier/frontend2/frontend_querier_queues.go:
```diff
@@ -0,0 +1,224 @@
+// Finds next queue for the querier. To support fair scheduling between users, client is expected
+// to pass last user index returned by this function as argument. Is there was no previous
+// last user index, use -1.
+func (q *queues) getNextQueueForQuerier(lastUserIndex int, querier string) (chan *schedulerRequest, string, int) {
```
It is cleaned by `*Scheduler. getNextRequestForQuerier` method. If there are no more requests for this user, user queue is deleted.

pstibrany commented on pkg/querier/frontend2/querier_scheduler_worker.go:
```diff
@@ -0,0 +1,434 @@
+
+	c, err := w.frontendPool.GetClientFor(frontendAddress)
+	if err == nil {
+		// Response is empty and uninteresting.
```
That comment refers to next line. Response from calling `QueryResult` on the frontend is empty structure.

pstibrany replied:
This is "happy path", we get here if we have something to report to frontend (can be query response or error response).

ranton256 replied:
Ah, I see, thanks.

pstibrany commented on pkg/querier/frontend2/scheduler.go:
```diff
@@ -0,0 +1,510 @@
+
+			// Ensure the request has not already expired.
+			if request.ctx.Err() == nil {
+				return request, lastUserIndex, nil
```
Not at this point. Deletion happens later in `forwardRequestToQuerier` via defered `cancelRequest` call.

ranton256 reviewed (approved):
LGTM, modulo the things you have called out as remaining TODOs.

csmarchbanks commented on CHANGELOG.md:
```diff
@@ -88,6 +88,7 @@
 * [ENHANCEMENT] Blocks storage: reduced number of bucket listing operations to list block content (applies to newly created blocks only). #3363
 * [ENHANCEMENT] Ruler: Include the tenant ID on the notifier logs. #3372
 * [ENHANCEMENT] Blocks storage Compactor: Added `-compactor.enabled-tenants` and `-compactor.disabled-tenants` to explicitly enable or disable compaction of specific tenants. #3385
+* [ENHANCEMENT] Query-Frontend / Query-Scheduler: New component called "Query-Scheduler" has been introduced. Query-Scheduler is simply a queue of requests, moved outside of Query-Frontend. This allows Query-Frontend to be scaled separately from number of queues. To make Query-Frontend and Querier use Query-Scheduler, they need to be started with `-frontend.scheduler-address` and `-querier.schedu ...
```
:wave: The v1.5 release just started and you will need to rebase to the latest master and move any CHANGELOG entries into the unreleased section. Thanks!

pracucci replied:
Done!

pracucci commented on pkg/querier/frontend2/scheduler.go:
```diff
@@ -0,0 +1,510 @@
+
+// getQueue picks a random queue and takes the next unexpired request off of it, so we
+// fairly process users queries.  Will block if there are no requests.
+func (s *Scheduler) getNextRequestForQuerier(ctx context.Context, lastUserIndex int, querierID string) (*schedulerRequest, int, error) {
```
This is also a copy of `Frontend.getNextRequestForQuerier()`

pstibrany replied:
Yes, I plan to refactor and unify this in follow-up PR, if that's fine. Idea is that frontend V1 will basically embed scheduler (hopefully while preserving its own metrics).

pracucci replied:
Ok, let's do it in a separate PR.

pracucci reviewed (commented):
Wow @pstibrany, such an huge work! A bit too big for my taste. Next time we should try to work through progressive PRs. It's too difficult to follow the changes and there's an high chance of let bugs slip in when such huge PRs come in.

That being said, solid work 👏 

I left a lot of minor comments and some questions. A part from this, would be nice seeing an integration test, at least to test the happy path.

Ideas to consider for future PRs:
- The query-frontend is a dedicated service in our architecture. We could make it a first class citizen in our package, moving it under `pkg/frontend`. If we'll need to keep v1/v2 code distinction, we could move version-specific code under `pkg/frontend/v1` and `pkg/frontend/v2` respectively, while keeping common code in `pkg/frontend`.

pstibrany commented:
> Wow @pstibrany, such an huge work! A bit too big for my taste. Next time we should try to work through progressive PRs. It's too difficult to follow the changes and there's an high chance of let bugs slip in when such huge PRs come in.

💯 agree. And with more time it would be possible to split this PR into multiple smaller ones, but given time constraints I opted for one big PR to get the work done asap.

Thank you for taking a time to review it!

> I left a lot of minor comments and some questions. A part from this, would be nice seeing an integration test, at least to test the happy path.

👍 

> * The query-frontend is a dedicated service in our architecture. We could make it a first class citizen in our package, moving it under `pkg/frontend`. If we'll need to keep v1/v2 code distinction, we could move version-specific code under `pkg/frontend/v1` and `pkg/frontend/v2` respectively, while keeping common code in `pkg/frontend`.

There is some repeated code between v1/v2, it might be possible to refactor it out and keep single frontend package. I plan to give it a try.

pstibrany commented on pkg/querier/frontend2/frontend2.go:
```diff
@@ -0,0 +1,290 @@
+	f.IntVar(&cfg.WorkerConcurrency, "frontend.scheduler-worker-concurrency", 5, "Number of goroutines pushing requests to ")
+
+	cfg.InfNames = []string{"eth0", "en0"}
+	f.Var((*flagext.StringSlice)(&cfg.InfNames), "frontend.interface", "Name of network interface to read address from.")
```
I took this from `lifecycler.interface`, where it is called just "interface", but I see we use `instance-interface-names` elsewhere, so that's what I am using here as well now. Thanks.

pstibrany commented on pkg/querier/frontend2/frontend2.go:
```diff
@@ -0,0 +1,290 @@
+
+// Config for a Frontend2.
+type Config struct {
+	SchedulerAddr     string                   `yaml:"scheduler_address"`
```
I've renamed this field (and one in `QuerierWorkersConfig`, but I'm not sure what you mean by "other places".

pracucci replied:
I meant `FrontendAddress`. Sorry for not being clear, and thanks!

pstibrany commented on pkg/querier/frontend/config.go:
```diff
@@ -0,0 +1,111 @@
+	cfg.WorkerV2.RegisterFlags(f)
+}
+
+func (cfg *CombinedWorkerConfig) InitQuerierWorker(querierCfg querier.Config, handler http.Handler, log log.Logger) (services.Service, error) {
```
I've removed this method from `*CombinedWorkerConfig` (and similarly from `*CombinedFrontendConfig`), but I'd prefer to keep the name, since `New...` indicates creation of new instance or error, but these functions can return `nil` result in some cases, which to me goes against `New...` naming. I also consider them to be extension of initialization done by init methods in `modules.go`.WDYT?

pracucci replied:
LGTM, thanks!

pstibrany commented on pkg/querier/frontend2/scheduler.go:
```diff
@@ -0,0 +1,510 @@
+)
+
+// Scheduler is responsible for queueing and dispatching queries to Queriers.
+type Scheduler struct {
```
That sounds like a good idea, but would prefer to do it in a follow-up PR, if that's fine?

Technically, we could also embed scheduler into Frontend V1, and cleanup the duplicate code between Frontend v1 and scheduler.

pracucci replied:
LGTM, thanks!

pstibrany commented on pkg/querier/frontend/downstream_roundtripper.go:
```diff
@@ -0,0 +1,40 @@
+		carrier := opentracing.HTTPHeadersCarrier(r.Header)
+		err := tracer.Inject(span.Context(), opentracing.HTTPHeaders, carrier)
+		if err != nil {
+			return nil, err
```
This can fail for these reasons:
- invalid span context is passed in first argument
- unknown format (`opentracing.HTTPHeaders`) passed as second argument
- invalid combination of format (`opentracing.HTTPHeaders`) and `carrier`

All of these are errors in the code, and cannot happen based on HTTP request itself. If code is correct, this call will never fail (it just sets some HTTP headers). On the other hand, ignoring the errors can mask possible bugs.

pstibrany commented on pkg/querier/frontend2/dns_watcher.go:
```diff
@@ -0,0 +1,80 @@
+// Notifications about address resolution. All notifications are sent on the same goroutine.
+type DNSNotifications interface {
+	AddressAdded(address string)
+
```
Added comments ;)

pstibrany commented on pkg/querier/frontend2/dns_watcher.go:
```diff
@@ -0,0 +1,80 @@
+			if servCtx.Err() != nil {
+				return nil
+			}
+			return errors.Wrapf(err, "error from DNS watcher")
```
Documentation for `Next` call says: `It should return an error if and only if Watcher cannot recover.`, so using backoff would not help.

Entire dns_watcher.go is copied and slightly modified from `pkg/querier/frontend/worker.go`, where it works in this way. This is also answer to your question about why it isn't based on Thanos DNS service discovery – I considered that, but decided against the idea in this PR, and just using what we have in `frontend` package already, mostly because then I would be reimplementing the watching service itself, which was not the point of this PR.

In a follow-up PR, my plan is to modify `pkg/querier/frontend/worker.go` to use new DNS watcher. We can then also change DNS watcher to use DNS service discovery from Thanos. I don't have strong opinion on it.

pracucci replied:
Thanks for the detailed explanation. All LGTM!

pstibrany commented on pkg/querier/frontend2/frontend.proto:
```diff
@@ -0,0 +1,94 @@
+}
+
+message SchedulerToQuerier {
+    uint64 queryID = 1;
```
> Could you add a comment to explain the queryID is required to be unique per querier

It is frontend that generates queryID, and each frontend has its own. Different frontends may generate the same queryID, and that is fine. Scheduler and querier don't care about this ID (scheduler uses it as a key, together with frontend address). Querier only passes it back to frontend.

I've added some comments explaining this (to `FrontendToScheduler` and `SchedulerToQuerier` messages).

pstibrany commented on pkg/querier/frontend2/frontend.proto:
```diff
@@ -0,0 +1,94 @@
+    rpc QueryResult (QueryResultRequest) returns (QueryResultResponse) { };
+}
+
+message QueryResultRequest {
```
In Cortex, userID in gRPC is mostly transferred via context (and call metadata). That isn't possible for long-running calls like `FrontendLoop` and `QuerierLoop`, because they are used for multiple users.

In case of `QueryResult` call however we *do* transfer userID in the context again. That's why it is not visible in the `QueryResultRequest`. I will add comment about it.

pstibrany commented on pkg/querier/frontend2/frontend2.go:
```diff
@@ -0,0 +1,290 @@
+		schedulerWorkers: schedulerWorkers,
+		requests:         newRequestsInProgress(),
+	}
+	// Randomize to avoid getting responses from queries sent before restart (which could lead to leak between tenants).
```
As explained in `QueryResultRequest` comment, frontend already knows userID from context. I've added the check – that's a good idea! Thanks!

pstibrany commented on pkg/querier/frontend2/frontend2.go:
```diff
@@ -0,0 +1,290 @@
+	}
+	// Randomize to avoid getting responses from queries sent before restart (which could lead to leak between tenants).
+	// This isn't perfect, but better than nothing.
+	f.lastQueryID.Store(rand.Uint64())
```
It will just wrap back to 0 and continue from there. No component really cares about actual values. Frontend only uses it for pairing requests and responses received from queriers.

pstibrany commented on pkg/querier/frontend2/frontend2.go:
```diff
@@ -0,0 +1,290 @@
+	tracer, span := opentracing.GlobalTracer(), opentracing.SpanFromContext(ctx)
+	if tracer != nil && span != nil {
+		carrier := (*httpgrpcHeadersCarrier)(req)
+		if err := tracer.Inject(span.Context(), opentracing.HTTPHeaders, carrier); err != nil {
```
Explained here: https://github.com/cortexproject/cortex/pull/3374#discussion_r514558011. WDYT?

pstibrany commented on pkg/querier/frontend2/frontend2.go:
```diff
@@ -0,0 +1,290 @@
+		case req.response <- qrReq.HttpResponse:
+			// Should always be possible, unless QueryResult is called multiple times with the same queryID.
+		default:
+			// If we cannot write to the channel, just ignore it.
```
This should not happen. Adding log line is a good idea.

pstibrany commented on pkg/querier/frontend2/frontend_scheduler_worker.go:
```diff
@@ -0,0 +1,303 @@
+		return
+	}
+
+	level.Debug(f.log).Log("msg", "adding connection to scheduler", "addr", address)
```
I left it as Debug because that's what `pkg/querier/frontend/worker.go` does. I agree it's useful, and I'm fine with using Info here, but then I would suggest changing `pkg/querier/frontend/worker.go` too. WDYT?

pracucci replied:
Let's do both in a separate PR.

pstibrany commented on pkg/querier/frontend2/frontend_scheduler_worker.go:
```diff
@@ -0,0 +1,303 @@
+		return
+	}
+
+	// If not, start a new one.
```
It was a continuation of previous comment (`// Already stopped or we already have worker for this address.`). Reworded.

pstibrany commented on pkg/querier/frontend2/frontend_scheduler_worker.go:
```diff
@@ -0,0 +1,303 @@
+}
+
+func (f *frontendSchedulerWorkers) AddressRemoved(address string) {
+	level.Debug(f.log).Log("msg", "removing connection to scheduler", "addr", address)
```
OK with change, but see my comment here: https://github.com/cortexproject/cortex/pull/3374#discussion_r514576859.

pstibrany commented on pkg/querier/frontend2/frontend_scheduler_worker.go:
```diff
@@ -0,0 +1,303 @@
+	// Can be nil if stopping has been called already.
+	if f.workers != nil {
+		f.workers[address] = w
+		go w.start()
```
This is just leftover from when `start` did more work before and was actually blocking. Fixed.

pstibrany commented on pkg/querier/frontend2/frontend_scheduler_worker.go:
```diff
@@ -0,0 +1,303 @@
+				return errors.New("scheduler is shutting down")
+
+			case ERROR:
+				req.enqueue <- enqueueResult{success: true, retry: false}
```
I've renamed `success` to `status` with two possible values: `wait_for_response` (tells frontend to wait for response, either from querier or frontend worker), or `failed_retry` which tells frontend that enqueuing has failed, and it should retry.

pstibrany replied:
It turns out all failures were with 'retry' flag, so it's just `wait_for_response` and `failed` now.

pstibrany commented on pkg/querier/frontend2/frontend_scheduler_worker.go:
```diff
@@ -0,0 +1,303 @@
+		if loopErr != nil {
+			level.Error(w.log).Log("msg", "error sending requests to scheduler", "err", loopErr, "addr", w.schedulerAddr)
+			backoff.Wait()
+			continue
```
Yes. It can restart with the same address, and worker should then reestablish connection and start new `FrontendLoop`.

If scheduler restarts with a different address, DNS watcher will call `AddressRemoved` which will eventually stop this method, and `AddressAdded` will then start a new loop.

pstibrany commented on pkg/querier/frontend2/dns_watcher.go:
```diff
@@ -0,0 +1,80 @@
+}
+
+// NewDNSWatcher creates a new DNS watcher and returns a service that is wrapping it.
+func NewDNSWatcher(address string, dnsLookupPeriod time.Duration, notifications DNSNotifications) (services.Service, error) {
```
Answered here: https://github.com/cortexproject/cortex/pull/3374#discussion_r514564131

pstibrany commented on pkg/querier/frontend2/querier_scheduler_worker.go:
```diff
@@ -0,0 +1,434 @@
+
+	frontendClientsGauge := promauto.With(reg).NewGauge(prometheus.GaugeOpts{
+		Name: "cortex_querier_scheduler_worker_clients",
+		Help: "The current number of frontend clients.",
```
Renamed to `cortex_querier_scheduler_worker_frontend_clients`. I wanted to keep consistent `cortex_querier_scheduler_worker` prefix, so it's a bit mouthful. Feel free to modify.

pracucci replied:
We're calling it query-scheduler everywhere. I'm going to rename all metrics from `cortex_querier_scheduler` to `cortex_query_scheduler`

pstibrany commented on pkg/querier/frontend2/querier_scheduler_worker.go:
```diff
@@ -0,0 +1,434 @@
+	}
+
+	level.Debug(f.log).Log("msg", "adding connection to scheduler", "addr", address)
+	conn, err := f.connectToScheduler(context.Background(), address)
```
`grpc.DialContext` is not blocking (unless explicitly called with `WithBlock()` option).

Note that this is also how pkg/querier/frontend/worker.go works.

pstibrany commented on pkg/querier/frontend2/frontend.proto:
```diff
@@ -0,0 +1,94 @@
+    // a "SchedulerToQuerier" messages containing HTTP requests and processes them. After processing the request,
+    // querier signals that it is ready to accept another one by sending empty QuerierToScheduler message.
+    //
+    // Long-running loop is used to detect broken connection between scheduler and querier. This is important
```
You're right it's missing currently.

pstibrany commented:
@pracucci Thanks a lot for your comprehensive review! I've addressed most of your feedback. In some cases GitHub UI doesn't allow me to add replies, so I've commented via Slack (feel free to share here if GH isn't misbehaving). I can address remaining feedback (integration test, grpc keep-alive) early next week, or feel free to take over if urgent.

Thanks!

pracucci reviewed (approved):
Thanks a lot @pstibrany for promptly addressing feedback! I've noted down (and shared to you) the list of remaining cleanup / enhancements / doc to do, so we can split the work and get this work "done done".
